[Feature] 유지보수 게시글 생성, 삭제, 수정을 구현한다.

ghrltjdtprbs opened the pull request:
# ⭐️ [Feature] 유지보수 게시글 생성, 삭제, 수정을 구현한다.

## 🛠️ 변경 사항

#### 유지보수 게시글 생성
- POST managers/repair/article

#### 유지보수 게시글 삭제
- DELETE managers/repair/article/{repairArticleId}
- 유지보수 게시글 삭제시 soft delete되며, 삭제 시점으로부터 일년이 지난 게시글과 이미지들은 스케줄러로 검사하여 데이터 베이스에서 자동 삭제됩니다.

#### 유지보수 게시글 수정
- PATCH managers/repair/article/{repairArticleId}
- 요구사항에는 나와있지 않지만, progress나 category 등을 단독으로도 간편하게 변경할 수 있도록 PATCH를 사용하여 수정 기능을 구현하였습니다.


## ❗️ 관련 이슈

- #68 

## ❗️ 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트

## ⏱️ 작업 기간

- 작업 시작일: (2024-05-22)
- 작업 종료일: (2024-05-23)

## ❓ 유의사항

- repairArticle 도메인에서 조회 기능도 같이 구현할 예정이라 관리자 cud는 ManagerRepairArticleController로 분리하였습니다. 후에 필요에 따라서 서비스 클래스도 분리할 예정입니다.

ghrltjdtprbs committed:
[feat] ManagerNotFoundException 예외 클래스 추가

변경 사항 :
 - ManagerNotFoundException 예외 클래스를 추가하였습니다.
 - ErrorCode에 MANAGER_NOT_FOUND를 추가하였습니다.

ghrltjdtprbs committed:
[feat] 유지보수 게시글 생성 기능 구현

변경 사항 :
 - 유지보수 게시글을 생성할 수 있는 기능을 구현하였습니다.

ghrltjdtprbs committed:
[refactor] RepairArticleImageDto 삭제

문제 사항 :
 - RepairArticleImageDto 클래스를 삭제하는 편이 더 가독성이 좋고 효율적임

변경 사항 :
 - RepairArticleImageDto 삭제 및 관련 코드 수정

ghrltjdtprbs committed:
[feat] 유지보수 게시글 삭제기능 추가

변경 사항 :
 - 게시글 작성자가 유지보수 게시글을 삭제할 수 있는 기능을 구현하였습니다.

ghrltjdtprbs committed:
[refactor] UNAUTHORIZED_ACCESS 의 http status 변경

변경 사항 :
 - UNAUTHORIZED_ACCESS 의 http status를 FORBIDDEN -> UNAUTHORIZED로 변경하였습니다.

ghrltjdtprbs committed:
[feat] soft delete된 유지보수 게시물 일괄 삭제기능 추가

문제 사항 :
 - soft delete된 게시물을 계속 가지고 있을 필요가 없음(db 용량 낭비)

변경 사항 :
 - scheduler을 이용해 삭제 시점으로부터 1년 이상된 게시물을 일괄 삭제하는 기능 구현
 - orphanRemoval = true 를 RepairArticle에 추가해 게시글이 삭제되면 관련 이미지도 자동 삭제되게 구현

ghrltjdtprbs committed:
[test] DeleteRepairArticleSchedulerTest 추가

문제 사항 :
 - 추가한 스케줄러가 실제로 동작하는지 테스트하기 위한 테스트 코드 추가

변경 사항 :
 - DeleteRepairArticleSchedulerTest 테스트 코드 추가

ghrltjdtprbs committed:
[feat] ArticleAlreadyDeletedException 예외 클래스 추가

변경 사항 :
 - 이미 삭제된 게시물에 작업을 시도할 경우 예외를 발생시키는 예외 클래스를 생성하였습니다.
 - 관련 메시지를 ErrorCode에 추가하였습니다.

ghrltjdtprbs committed:
[feat] 게시글 수정 controller 추가

변경 사항 :
 - 게시글 수정 controller를 추가하였습니다.

ghrltjdtprbs committed:
[refactor] CreateRepairArticleRequestDto 검증 추가

변경 사항 :
 - 게시글 생성에 사용되는 dto에 검증을 추가하였습니다.

ghrltjdtprbs committed:
[feat] 게시글 수정 로직 추가

변경 사항 :
 - 게시글 수정 관련 dto,service로직을 추가하였습니다.

ghrltjdtprbs committed:
[feat] 게시글 수정 controller에 @Valid삭제

변경 사항 :
 - 게시글 수정 dto에서는 검증이 필요 없어 @Valid를 제거하였습니다.

ghrltjdtprbs committed:
[test] DeleteRepairArticleSchedulerTest 추가

변경 사항 :
 - DeleteRepairArticleScheduler 가 잘 동작하는지 테스트하기 위한 테스트 코드를 추가하였습니다.

ghrltjdtprbs committed:
[refactor] controller 로그인 방식 변경

문제 사항 :
 - 프로젝트의 로그인 방식이 변경되어 controller를 수정하였습니다.

Galmaeki commented on src/main/java/com/final_10aeat/common/scheduler/config/SchedulerConfig.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.scheduling.annotation.EnableScheduling;
+
+@Configuration
+@EnableScheduling
```
Config에 구현없이 어노테이션만 붙은것들은 합쳐도 좋을것 같습니다!

ghrltjdtprbs replied:
후에 스케줄링 관련 설정이 더 추가될 예정입니다! 
현재 SchedulerConfig 클래스에 구현 부분이 없지만, 확장성을 고려해서 추가했습니다! 👍👍

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/controller/ManagerRepairArticleController.java:
```diff
@@ -0,0 +1,58 @@
+@RestController
+@RequiredArgsConstructor
+@PreAuthorize("hasRole('MANAGER')")
+@RequestMapping("/managers/repair/article")
```
article을 articles로 사용하는건 어떨까요?

ghrltjdtprbs replied:
반영해서 수정했습니다! 

관련 커밋 : ca2f3c9f98d888212b3f067b59abb3c9ed3f23ba

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/controller/ManagerRepairArticleController.java:
```diff
@@ -0,0 +1,58 @@
+    private final RepairArticleService repairArticleService;
+
+    @PostMapping
+    public ResponseEntity<ResponseDTO<Void>> createRepairArticle(@RequestBody @Valid
```
create로만 사용하는건 어떻게 생각하시나요?
class 명에도 RepairArticleController.createRepairArticle 메소드라서  줄여도 좋을것 같슴다

ghrltjdtprbs replied:
후에 안건 진행상황 생성, 수정도 해당 서비스 클래스에 붙일 예정이라 구분을 쉽게 하기 위해 createRepairArticle로 명시하였습니다! 😊

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/service/RepairArticleService.java:
```diff
@@ -0,0 +1,126 @@
+
+    public void createRepairArticle(CreateRepairArticleRequestDto request,
+        Long managerId) {
+        Manager manager = managerRepository.findById(managerId).orElseThrow(
```
작성자의 id를 담당 매니저로 사용하는 구조인데 Controller에서 Manager을 바로 사용하지 않으신 이유가 있을까요?

ghrltjdtprbs replied:
1. Controller와 Service계층의 역할을 명확하게 하기 위함 
2. Manager 객체를 직접 다루지 않고 ID를 통해 검증하면서 보안성을 강화 및 잘못된 사용자 정보의 처리 리스크를 최소화

이 정도의 이유가 있는데, 사실 크게 생각한 부분은 아닙니다! 🥲

Galmaeki commented on src/main/java/com/final_10aeat/domain/repairArticle/dto/request/UpdateRepairArticleRequestDto.java:
```diff
@@ -0,0 +1,20 @@
+    LocalDateTime constructionEnd,
+    String repairCompany,
+    String repairCompanyWebsite,
+    List<String> images
```
이렇게해도 이미지가 오는건가요?
아니면 임시 처리인가요?

ghrltjdtprbs replied:
리스트로 받은 이미지들이 repair_image테이블에 각각 저장됩니다!

Galmaeki replied:
MultipartFile로 안해도 이미지가 String로 변환돼서 들어오는건가요?

ghrltjdtprbs replied:
이미지가 외부 서비스에 미리 업로드 되어 있고, 해당 URL을 통해 접근 가능하다는 가정 하에 설계된 코드입니다! 현재 구현에서는 이미지를 직접 파일 형태로 업로드 받는 대신, 이미지의 URL을 String 리스트로 받고 있어요!

Galmaeki reviewed (changes requested):
고생 많으셨습니다!

ghrltjdtprbs committed:
[refactor] 유지보수 cud경로 변경

문제 사항 :
 - 코드리뷰 반영

변경 사항 :
 - /managers/repair/article -> /managers/repair/articles

asd42270 reviewed (approved):
오 접근 권한 예외처리까지 대박... 고생하셨습니다!!
